### PR TITLE
Run nightly tests against Spark 3

### PR DESCRIPTION
**Pull request by karenfeng**
## What changes are proposed in this pull request?

Simplifies the workflow to build/test on different versions on Spark, so the developer only needs to modify the environment variable `SPARK_VERSION`.

Also creates a per-job and nightly trigger to make sure we're compatible with the nightly Spark 3 snapshot release, and the most recent PySpark release (dev2).

Related to https://github.com/projectglow/glow/issues/172.

## How is this patch tested?
- [x] Unit tests
- [x] Integration tests
- [ ] Manual tests

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -131,10 +131,45 @@ jobs:
...
+          keys:
+            - conda-deps-v1-{{ checksum "python/environment.yml" }}
+      - *install_deps
+      - run:
+          name: Change Spark version
+          environment:
+          command: |
+            echo "3.0.0-SNAPSHOT" > spark-version.txt
+      - run:
+          name: Run Scala tests
```
Comment (henrydavidge) [inline .circleci/config.yml:148]: Seems like we should bump the pyspark version and run Python tests as well

File: .circleci/config.yml [hunk 2]
```diff
@@ -131,10 +131,45 @@ jobs:
...
+      - checkout
+      - restore_cache:
+          keys:
+            - conda-deps-v1-{{ checksum "python/environment.yml" }}
+      - *install_deps
+      - run:
+          name: Change Spark version
+          environment:
+          command: |
+            echo "3.0.0-SNAPSHOT" > spark-version.txt
```
Comment (henrydavidge) [inline .circleci/config.yml:146]: Why read this from a text file? I think it's more standard to read from an environment variable or a system property. Then there's less indirection.

File: .circleci/config.yml (hunk 1 above)
Comment (karenfeng) [reply .circleci/config.yml:148]: PySpark isn't released in the nightly snapshots. If we wanted to test against PySpark 3.0, we'd have to use one of the preview releases (eg. https://archive.apache.org/dist/spark/spark-3.0.0-preview2/), which we are currently ahead of.

File: .circleci/config.yml (hunk 2 above)
Comment (karenfeng) [reply .circleci/config.yml]: Sure, I made the change.

File: .circleci/config.yml [hunk 3]
```diff
@@ -131,10 +131,58 @@ jobs:
...
+      - restore_cache:
+          keys:
+            - conda-deps-v1-{{ checksum "python/environment.yml" }}
+      - *install_deps
+      - run:
+          name: Use latest PySpark release
+          environment:
+          command: |
+            export PATH=$HOME/conda/envs/glow/bin:$PATH
+            wget https://archive.apache.org/dist/spark/spark-3.0.0-preview2/pyspark-3.0.0.dev2.tar.gz
```
Comment (henrydavidge) [inline .circleci/config.yml:147]: Doesn't this mess up the cached conda environment?

File: .circleci/config.yml [hunk 4]
```diff
@@ -131,10 +131,58 @@ jobs:
...
+            sbt docs_2_12/test exit
+
 workflows:
   version: 2
   test:
     jobs:
       - check-links
       - scala-2_11-tests
       - scala-2_12-tests
+      - spark-3-tests
```
Comment (henrydavidge) [inline .circleci/config.yml:178]: So we want to run this build against PRs and nightly?

**Comment by Hoeze** [comment]
Is there already a working build for Spark 3.0-rc1?

**Comment by karenfeng** [comment]
@Hoeze, our build is compatible with the nightly Spark 3 snapshots; this might be ahead of rc1. We don't plan to release Glow on Spark 3 until after the public release, but in the meantime you can build a fat jar locally with https://github.com/sbt/sbt-assembly.

File: build.sbt [hunk 5]
```diff
@@ -1,15 +1,21 @@
...
 
-val sparkVersionMajorMinor = "2.4"
-val sparkVersion = "2.4.3"
-
 lazy val scala212 = "2.12.8"
 lazy val scala211 = "2.11.12"
 
+lazy val sparkVersion = sys.env.getOrElse("SPARK_VERSION", "2.4.3")
+
+lazy val minorVersion = (version: String) =>
```
Comment (henrydavidge) [inline build.sbt:13]: nit: this should still be called `majorMinor`. Also, why is it a lazy val pointing to an anonymous function? Can it just be a lazy val, or at least a def?

**Pull request merged**